### PR TITLE
Allow runners to really use runner.nthreads, code cleanup.

### PR DESCRIPTION
**Pull request by ftence**
For a detailled explanation on ThreadPoolExecutor, see:
http://www.bigsoft.co.uk/blog/2009/11/27/rules-of-a-threadpoolexecutor-pool-size

### CODE REVIEW

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 1]
```diff
@@ -39,265 +68,226 @@
...
-      this.nthreads = Integer.parseInt(config.getProperty(Configuration.RUNNER_NTHREADS));
+      int nthreads = Integer.parseInt(config.getProperty(Configuration.RUNNER_NTHREADS));
       this.scanperiod = Long.parseLong(config.getProperty(Configuration.RUNNER_SCANPERIOD));
       this.minperiod = Long.parseLong(config.getProperty(Configuration.RUNNER_MINPERIOD));
       this.endpoint = config.getProperty(Configuration.RUNNER_ENDPOINT);
-      
-      this.executor = new ThreadPoolExecutor(1, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256));
-  
+
+      ThreadPoolExecutor runnersExecutor = new ThreadPoolExecutor(nthreads, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256), new ThreadFactory() {
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Here is the important change.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 2]
```diff
@@ -252,7 +254,11 @@ public Thread newThread(Runnable r) {
       String strategy = config.getProperty(Configuration.RUNNER_KAFKA_CONSUMER_PARTITION_ASSIGNMENT_STRATEGY);
       int nthreads = Integer.parseInt(config.getProperty(Configuration.RUNNER_KAFKA_NTHREADS));
       long commitPeriod = Long.parseLong(config.getProperty(Configuration.RUNNER_KAFKA_COMMITPERIOD));
-      this.executor = new ThreadPoolExecutor(1, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256));
+
+      ThreadPoolExecutor runnersExecutor = new ThreadPoolExecutor(nthreads, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256), new NamedThreadFactory());
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Here is the important change.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 3]
```diff
@@ -39,265 +68,232 @@
...
-    } else {
-      isWorker = false;
-    }
-  
-    if (isStandalone || isScheduler) {      
+    isStandalone = configuredRoles.contains("standalone");
+
+    isScheduler = configuredRoles.contains("scheduler");
+
+    boolean isWorker = configuredRoles.contains("worker");
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Why redefining isWorker here?

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 4]
```diff
@@ -39,265 +68,232 @@
...
-      this.executor = new ThreadPoolExecutor(1, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256));
-      
-      this.consumerPool = new KafkaSynchronizedConsumerPool(zkconnect, topic, clientid, groupid, strategy, nthreads, commitPeriod, new ScriptRunnerConsumerFactory(this));
+
+      ThreadPoolExecutor runnersExecutor = new ThreadPoolExecutor(nthreads, nthreads, 30000L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>(nthreads * 256), new NamedThreadFactory());
+      runnersExecutor.allowCoreThreadTimeOut(true);
+
+      this.executor = runnersExecutor;
+
+      new KafkaSynchronizedConsumerPool(zkconnect, topic, clientid, groupid, strategy, nthreads, commitPeriod, new ScriptRunnerConsumerFactory(this));
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Code is clearer if you keep a reference to the created consumer pool

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 5]
```diff
@@ -39,265 +68,232 @@
...
  */
 public class ScriptRunner extends Thread {
-  
+
+  private class NamedThreadFactory implements ThreadFactory {
+    private AtomicLong threadNumber = new AtomicLong();
+
+    @Override
+    public Thread newThread(Runnable r) {
+      return new Thread(r, "[Warp ScriptRunner Executor #" + threadNumber.getAndIncrement() + "]");
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Thread needs to be a daemon Thread.
Group needs to be set.
Priority needs to be set.

Check the DefaultThreadFactory from Executor.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java (hunk 3 above)
Comment (ftence) [reply warp10/src/main/java/io/warp10/script/ScriptRunner.java]: isWorker was a member of ScriptRunner but was only used in its constructor so I made it a local variable.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java (hunk 4 above)
Comment (ftence) [reply warp10/src/main/java/io/warp10/script/ScriptRunner.java]: I do agree this line looks strange without keeping a reference but not storing it in a member of the class makes it clear that it is not used/modified elsewhere.
Can a local variable do? What about the GC, could it destroy the instance if there are no ref to this instance?

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java (hunk 5 above)
Comment (ftence) [reply warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Thanks for the pointer, will fix.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java (hunk 3 above)
Comment (hbs) [reply warp10/src/main/java/io/warp10/script/ScriptRunner.java]: The overall understanding of the code is probably better if isWorker is defined with the other two, even if it is not used (yet?).

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java (hunk 4 above)
Comment (hbs) [reply warp10/src/main/java/io/warp10/script/ScriptRunner.java]: The GC issue would have to be investigated, a local variable would not work if the GC can discard the object, so for now it's probably safer to stick with the setting of the field.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 6]
```diff
@@ -39,265 +68,248 @@
...
 public class ScriptRunner extends Thread {
-  
+
+  static class NamedThreadFactory implements ThreadFactory {
+    final ThreadGroup group;
+    final AtomicInteger threadNumber = new AtomicInteger(1);
+
+    NamedThreadFactory() {
+      SecurityManager s = System.getSecurityManager();
+      group = (s != null) ? s.getThreadGroup() :
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: leave on a single line.

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 7]
```diff
@@ -39,265 +68,248 @@
...
+          Thread.currentThread().getThreadGroup();
+    }
+
+    public Thread newThread(Runnable r) {
+      Thread t = new Thread(group, r,
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Leave on a single line

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 8]
```diff
@@ -39,265 +68,248 @@
...
+          "[Warp ScriptRunner Thread #" + threadNumber.getAndIncrement() + "]",
+          0);
+      if (t.isDaemon())
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Use braces {}

File: warp10/src/main/java/io/warp10/script/ScriptRunner.java [hunk 9]
```diff
@@ -39,265 +68,248 @@
...
+        t.setDaemon(false);
+      if (t.getPriority() != Thread.NORM_PRIORITY)
```
Comment (hbs) [inline warp10/src/main/java/io/warp10/script/ScriptRunner.java]: Use braces {}

**Pull request merged**